tests/test_server.py

```diff
@@ -52,140 +71,152 @@ def test_add_device(client, name='steelseries-rival310'):
     data = {
         'shortname': name
     }
-    response = client.post('/devices/add',
-                           data=json.dumps(data),
-                           content_type='application/json')
+    response = client.post('/devices/add', json=data)
     assert response.status_code == 201
-    data = json.loads(response.data)
+    data = response.json()
     assert 'id' in data
     assert 'shortname' in data
     assert 'name' in data
-    return data['id']
 
-
-def test_get_device(client):
-    id = test_add_device(client)
-
-    response = client.get('/devices/{}'.format(id))
-    assert response.status_code == 200
-    data = json.loads(response.data)
-    assert 'id' in data
-    assert 'shortname' in data
-    assert 'name' in data
+    client.delete('/devices/{}'.format(data['id']))
 
 
 def test_delete_device(client):
-    id = test_add_device(client)
+    id = add_device(client)
 
     response = client.delete('/devices/{}'.format(id))
     assert response.status_code == 204
 
 
-def test_get_dpi(client, dpi_id=0):
-    from ratbag_emu.device_handler import DeviceHandler
-
-    id = test_add_device(client)
+def test_get_device(client):
+    id = add_device(client)
 
-    response = client.get('/devices/{}/phys_props/dpi/{}'.format(id, dpi_id))
+    response = client.get('/devices/{}'.format(id))
     assert response.status_code == 200
-    data = json.loads(response.data)
-    assert data == DeviceHandler.devices[id].hprof.dpi[dpi_id if
-        dpi_id != 'active' else DeviceHandler.devices[id].hprof.active_dpi]
+    data = response.json()
+    assert 'id' in data
+    assert 'shortname' in data
+    assert 'name' in data
+    assert 'input_nodes' in data
```

Comment:
python's unittest has a bunch of helpers like [`TestCase.assertIn`](https://docs.python.org/3/library/unittest.html) which has the added benefit of having useful debug output when it fails.

Reply:
Does it really make sense to have a class here? I dislike that approach.

Reply:
why? the only noticeable extra work you have is having to move everything right by one indent. benefits are better encapsulation and the ability to re-test things by subclassing. If a test class subclasses another class it will run all tests from the parent, so it's easy to stack tests that way. Every single time I wrote unittests the classes have come in handy.

and the `assertIn` helpers have the advantage of printing the list contents so you can immediately see what else is in that list. `assert` on its own is only useful for "this really mustn't happen" in normal code.